pkg/controller/control.go

```diff
@@ -901,3 +917,117 @@ func (c *Controller) BackupReplicaMappingDelete(id string) error {
+		return err
+	} else if mountPoint == "" {
+		logrus.WithField("volume", volume).Info("Volume is not mounted, no need to unfreeze filesystem")
+		return nil
+	}
+
+	if err = unfreezeFilesystem(ne, mountPoint); err != nil {
+		logrus.WithField("volume", volume).WithError(err).Errorf("Failed to unfreeze the filesystem at %v", mountPoint)
+		return err
+	}
+
+	logrus.WithField("volume", volume).Infof("Unfroze filesystem of volume mounted at %v", mountPoint)
+	return nil
+}
+
+func findMountPoint(ne *iutil.NamespaceExecutor, devicePath string) (string, error) {
+	/* Example command output
+	   # all of these run with nsenter --mount=/host/proc/1/ns/mnt --net=/host/proc/1/ns/net
+	   # found mount, returns 0
+	   $ findmnt -o TARGET,FSTYPE --noheadings --first-only --source /dev/longhorn/pvc-397cfa17-70d1-4780-bac7-bbff7e00d7ec
+	   /var/lib/kubelet/pods/4d097f01-bb44-4d7d-8905-a9c59ff09468/volumes/kubernetes.io~csi/pvc-397cfa17-70d1-4780-bac7-bbff7e00d7ec/mount ext4
+
+	   # failed to find mount, returns 1
+	   $ findmnt /dev/longhorn/unknown
+	*/
+
+	logrus.Debugf("Searching for mount point of device %v", devicePath)
+	var ee *exec.ExitError
+	out, err := ne.Execute("findmnt", []string{"-o", "target,fstype", "--first-only", "--noheadings", "--source", devicePath})
+	if err != nil && errors.As(err, &ee) && ee.ExitCode() == 1 {
+		// findmnt return error code 1 if no mount is found
+		// so we need to unwrap the returned error to get the exit code
+		logrus.WithError(ee).Debugf("Device %v is not mounted", devicePath)
+		return "", nil
+	}
+
+	var mountPoint string
+	if i := strings.LastIndex(strings.TrimSuffix(out, "\n"), " "); i != -1 {
+		mountPoint = out[:i]
+		logrus.Debugf("Found mount point %v for device %v", mountPoint, devicePath)
+	}
+
+	return mountPoint, err
+}
+
+func freezeFilesystem(ne *iutil.NamespaceExecutor, mountPoint string) error {
+	/* Example Command output
+	   # successful freeze, returns 0, no output
+	   $ fsfreeze -f /var/lib/kubelet/pods/4d097f01-bb44-4d7d-8905-a9c59ff09468/volumes/kubernetes.io~csi/pvc-397cfa17-70d1-4780-bac7-bbff7e00d7ec/mount
+
+	   # already frozen, returns 1
+	   $ fsfreeze -f /var/lib/kubelet/pods/4d097f01-bb44-4d7d-8905-a9c59ff09468/volumes/kubernetes.io~csi/pvc-397cfa17-70d1-4780-bac7-bbff7e00d7ec/mount
+	   fsfreeze: /var/lib/kubelet/pods/4d097f01-bb44-4d7d-8905-a9c59ff09468/volumes/kubernetes.io~csi/pvc-397cfa17-70d1-4780-bac7-bbff7e00d7ec/mount: freeze failed: Device or resource busy
+	*/
+	var ee *exec.ExitError
+	_, err := ne.ExecuteWithTimeout(freezeTimeout, "fsfreeze", []string{"-f", mountPoint})
+	if err != nil && errors.As(err, &ee) && ee.ExitCode() == 1 {
+		// most likely the filesystem is already frozen or the mount point doesn't exist
+		logrus.WithError(err).Debugf("Freeze of filesystem mounted at %v failed", mountPoint)
+		return nil
+	}
+
+	return err
+}
+
+func unfreezeFilesystem(ne *iutil.NamespaceExecutor, mountPoint string) error {
```

Comment:
if `unfreeze` failed due to some reasons other than invalid parameters/path, will it affect the volume io?

Reply:
If a frozen filesystem does not get unfrozen then it remains frozen.
Deferred functions will be called even in the presence of a panic.

Reply:
I mean, should Longhorn return an error and mark the volume as failed if the unfreezing failed? Then Longhorn or users would restart or clean up it.

Reply:
There isn't really a good way to tell whether a filesystem is frozen. So it's naturally that unfreeze will fail in the case where the filesystem is not frozen. Unfreezing should not fail for a frozen filesystem. If the volume gets detached from the node and attached again the filesystem will be in an unfrozen state, even if previously frozen.

So in the case where the node crashed after the freeze but before the unfreeze, the volume will be detached and reattached on a new node which will mean the filesystem is unfrozen.